File: packages/protocol/src/transfer-loopback.test.ts
import { describe, it, expect } from 'vitest';
import { createHash } from 'node:crypto';
import { deriveTransferKeys } from './keyschedule.js';
import { bytesSource, MemorySink, type Destination, type Digest } from './transfer-ports.js';
import { FrameType } from './transfer.js';
import type { FileEntry, Manifest } from './transfer.js';
import { TransferSender } from './transfer-sender.js';
import { TransferReceiver } from './transfer-receiver.js';

function nodeDigest(): Digest {
  const h = createHash('sha256');
  return { update: (b) => void h.update(b), hexDigest: () => h.digest('hex') };
}
const master = new Uint8Array(32).fill(3);

/** Wire a sender (offerer role, o2j) to a receiver (joiner role) via async queues. */
async function runTransfer(
  data: Uint8Array,
  opts: { blockSize: number; frameSize: number; window: number; corrupt?: boolean },
) {
  const keys = await deriveTransferKeys(master);
  const sink = new MemorySink();

  // The two engines reference each other; a holder breaks the declaration cycle.
  const box: { receiver?: TransferReceiver } = {};
  const sender = new TransferSender({
    file: bytesSource(data, {
      name: 'f',
      size: data.length,
      mime: 'application/octet-stream',
      lastModified: 1,
    }),
    send: (f) => queueMicrotask(() => box.receiver!.handle(f)),
    sendDir: keys.o2j,
    recvDir: keys.j2o,
    sendCounterStart: 0,
    recvCounterStart: 0,
    createDigest: nodeDigest,
    blockSize: opts.blockSize,
    frameSize: opts.frameSize,
    window: opts.window,
  });
  let corrupted = false;
  const receiver = new TransferReceiver({
    send: (f) => queueMicrotask(() => sender.handle(f)),
    sendDir: keys.j2o,
    recvDir: keys.o2j,
    sendCounterStart: 0,
    recvCounterStart: 0,
    createDigest: nodeDigest,
    sink,
  });
  box.receiver = receiver;

  if (opts.corrupt) {
    // Re-wrap send to flip one byte in the second frame (a block_data payload/tag).
    let n = 0;
    (sender as unknown as { o: { send: (f: Uint8Array) => void } }).o.send = (f: Uint8Array) => {
      if (!corrupted && n++ === 1) {
        f = f.slice();
        f[f.length - 1] ^= 0x01;
        corrupted = true;
      }
      queueMicrotask(() => receiver.handle(f));
    };
  }

  const runP = sender.run();
  const result = await Promise.allSettled([runP, receiver.done]);
  return { result, sink };
}

describe('transfer loopback (no browser)', () => {
  const sizes = [0, 1, 7, 8, 20, 4096, 100_000];
  for (const size of sizes) {
    it(`streams ${size} bytes end-to-end with a matching digest`, async () => {
      const data = new Uint8Array(size).map((_, i) => (i * 131 + 7) & 0xff);
      const { result, sink } = await runTransfer(data, {
        blockSize: 1024,
        frameSize: 256,
        window: 4,
      });
      expect(result[0].status).toBe('fulfilled');
      expect(result[1].status).toBe('fulfilled');
      expect([...sink.bytes()]).toEqual([...data]);
      if (result[1].status === 'fulfilled') {
        expect(result[1].value.digest).toBe(createHash('sha256').update(data).digest('hex'));
      }
    });
  }

  it('aborts loudly when a frame is corrupted in flight', async () => {
    const data = new Uint8Array(4096).map((_, i) => i & 0xff);
    const { result, sink } = await runTransfer(data, {
      blockSize: 1024,
      frameSize: 256,
      window: 4,
      corrupt: true,
    });
    expect(result.some((r) => r.status === 'rejected')).toBe(true);
    expect(sink.isClosed).toBe(false);
  });

  it('re-sends Done when a cutover drops it after the receiver settled', async () => {
    // The receiver settles and sends its one-shot Done, but that Done is lost — the exact
    // loss a direct→relay cutover can cause right after every block is acknowledged. The
    // sender retransmits Complete on the new path; a settled receiver must re-answer with
    // a fresh Done or the sender stalls waiting for it.
    const keys = await deriveTransferKeys(master);
    const data = new Uint8Array(64_000).map((_, i) => (i * 131 + 7) & 0xff);
    const sink = new MemorySink();

    const box: { receiver?: TransferReceiver } = {};
    let didCutover = false;
    let doneDropped = false;
    const sender = new TransferSender({
      file: bytesSource(data, {
        name: 'f',
        size: data.length,
        mime: 'application/octet-stream',
        lastModified: 1,
      }),
      send: (frame) => queueMicrotask(() => box.receiver!.handle(frame)),
      sendDir: keys.o2j,
      recvDir: keys.j2o,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      blockSize: 1024,
      frameSize: 256,
      window: 4,
      doneTimeoutMs: 2000, // keep the no-fix failure fast
    });
    const receiver = new TransferReceiver({
      send: (frame) => {
        queueMicrotask(() => {
          // Drop the receiver's first Done (type byte at sealed offset 9), then simulate
          // the sender's path change so it retransmits Complete.
          if (!doneDropped && frame[9] === FrameType.Done) {
            doneDropped = true;
            if (!didCutover) {
              didCutover = true;
              sender.transportChanged();
            }
            return;
          }
          sender.handle(frame);
        });
      },
      sendDir: keys.j2o,
      recvDir: keys.o2j,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      sink,
    });
    box.receiver = receiver;

    const [digest, result] = await Promise.all([sender.run(), receiver.done]);
    expect(doneDropped).toBe(true);
    expect([...sink.bytes()]).toEqual([...data]);
    expect(result.digest).toBe(createHash('sha256').update(data).digest('hex'));
    expect(digest).toBe(result.digest);
  });

  it('streams a nested multi-file set with empty files and aggregate progress', async () => {
    const keys = await deriveTransferKeys(master);
    const contents = [
      new Uint8Array([1, 2, 3]),
      new Uint8Array(),
      new Uint8Array(3000).map((_, i) => i & 0xff),
    ];
    const names = ['folder/a.bin', 'folder/empty.txt', 'b.bin'];
    const sinks = new Map<string, MemorySink>();
    let prepared: Manifest | undefined;
    let committed = false;
    const destination: Destination = {
      prepare(manifest) {
        prepared = manifest;
      },
      open(file: FileEntry) {
        const sink = new MemorySink();
        sinks.set(file.name, sink);
        return sink;
      },
      close() {
        committed = true;
      },
      abort(reason) {
        for (const sink of sinks.values()) sink.abort(reason);
      },
    };
    const box: { receiver?: TransferReceiver } = {};
    const progress: number[] = [];
    const sender = new TransferSender({
      files: contents.map((bytes, idx) =>
        bytesSource(bytes, { name: names[idx]!, size: bytes.length, mime: '', lastModified: 1 }),
      ),
      send: (frame) => queueMicrotask(() => box.receiver!.handle(frame)),
      sendDir: keys.o2j,
      recvDir: keys.j2o,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      blockSize: 1024,
      frameSize: 256,
      window: 2,
      onProgress: (bytes) => progress.push(bytes),
    });
    const receiver = new TransferReceiver({
      send: (frame) => queueMicrotask(() => sender.handle(frame)),
      sendDir: keys.j2o,
      recvDir: keys.o2j,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      destination,
    });
    box.receiver = receiver;

    const [sendDigest, received] = await Promise.all([sender.run(), receiver.done]);
    expect(prepared?.files.map((file) => file.name)).toEqual(names);
    expect(received.files.map((file) => file.name)).toEqual(names);
    expect(received.digest).toBe(sendDigest);
    expect(received.digests).toEqual(
      contents.map((bytes) => createHash('sha256').update(bytes).digest('hex')),
    );
    for (const [idx, name] of names.entries())
      expect(sinks.get(name)?.bytes()).toEqual(contents[idx]);
    expect(progress.at(-1)).toBe(contents.reduce((total, bytes) => total + bytes.length, 0));
    expect(committed).toBe(true);
  });

  it('resumes a partially received transfer from the committed high-water mark', async () => {
    const keys = await deriveTransferKeys(master);
    const blockSize = 1024;
    const data = new Uint8Array(100_000).map((_, i) => (i * 131 + 7) & 0xff);
    const blocks = Math.ceil(data.length / blockSize);
    const stop = 10;
    const prefixLen = stop * blockSize;
    const transferId = 'a'.repeat(32);

    // Model a reloaded receiver: the first `stop` blocks are already persisted, so restore a
    // prefilled sink and a digest already fed exactly those bytes.
    const sink = new MemorySink();
    sink.write(0, data.subarray(0, prefixLen));
    const seed = nodeDigest();
    seed.update(data.subarray(0, prefixLen));

    let commits = 0;
    const box: { receiver?: TransferReceiver } = {};
    const sender = new TransferSender({
      file: bytesSource(data, { name: 'f', size: data.length, mime: '', lastModified: 1 }),
      send: (f) => queueMicrotask(() => box.receiver!.handle(f)),
      sendDir: keys.o2j,
      recvDir: keys.j2o,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      blockSize,
      frameSize: 256,
      window: 4,
      transferId,
    });
    const receiver = new TransferReceiver({
      send: (f) => queueMicrotask(() => sender.handle(f)),
      sendDir: keys.j2o,
      recvDir: keys.o2j,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      sink,
      resume: { transferId, files: new Map([[0, { haveBlocks: stop, seedDigest: seed }]]) },
      onProgress: () => void commits++,
    });
    box.receiver = receiver;

    const [digest, result] = await Promise.all([sender.run(), receiver.done]);
    expect([...sink.bytes()]).toEqual([...data]);
    expect(result.digest).toBe(createHash('sha256').update(data).digest('hex'));
    expect(digest).toBe(result.digest);
    // Only the blocks the receiver lacked are streamed and committed on resume.
    expect(commits).toBe(blocks - stop);
  });

  it('ignores a resume seed whose transferId does not match the manifest', async () => {
    const keys = await deriveTransferKeys(master);
    const blockSize = 1024;
    const data = new Uint8Array(20_000).map((_, i) => (i * 131 + 7) & 0xff);
    const blocks = Math.ceil(data.length / blockSize);

    // A seed for a *different* transfer must be discarded: wrong high-water and a poisoned digest.
    // If it were wrongly applied, the whole-file digest would fail and receiver.done would reject.
    const staleSeed = nodeDigest();
    staleSeed.update(new Uint8Array(blockSize).fill(0xff));

    let commits = 0;
    const sink = new MemorySink();
    const box: { receiver?: TransferReceiver } = {};
    const sender = new TransferSender({
      file: bytesSource(data, { name: 'f', size: data.length, mime: '', lastModified: 1 }),
      send: (f) => queueMicrotask(() => box.receiver!.handle(f)),
      sendDir: keys.o2j,
      recvDir: keys.j2o,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      blockSize,
      frameSize: 256,
      window: 4,
      transferId: 'b'.repeat(32),
    });
    const receiver = new TransferReceiver({
      send: (f) => queueMicrotask(() => sender.handle(f)),
      sendDir: keys.j2o,
      recvDir: keys.o2j,
      sendCounterStart: 0,
      recvCounterStart: 0,
      createDigest: nodeDigest,
      sink,
      resume: {
        transferId: 'a'.repeat(32),
        files: new Map([[0, { haveBlocks: 1, seedDigest: staleSeed }]]),
      },
      onProgress: () => void commits++,
    });
    box.receiver = receiver;

    const [digest, result] = await Promise.all([sender.run(), receiver.done]);
    expect([...sink.bytes()]).toEqual([...data]);
    expect(result.digest).toBe(createHash('sha256').update(data).digest('hex'));
    expect(digest).toBe(result.digest);
    // Fresh receive: every block is streamed because the stale seed was rejected.
    expect(commits).toBe(blocks);
  });
});

File: packages/wire/transfer_loopback_test.go
package wire

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"sync/atomic"
	"testing"
	"time"
)

// loopbackMaster is the fixed master key the loopback derives both directions from.
func loopbackMaster() []byte {
	m := make([]byte, 32)
	for i := range m {
		m[i] = 3
	}
	return m
}

type loopbackResult struct {
	runErr  error
	recvRes ReceiveResult
	recvErr error
	sink    *MemorySink
}

// runLoopback wires a Sender (offerer, o2j) to a Receiver (joiner) over two buffered channels,
// each drained by its own goroutine so delivery is ordered and genuinely concurrent — the same
// shape as a real reliable, ordered DataChannel. When corrupt is set, one byte of the second
// frame the sender emits is flipped in flight, forcing an integrity abort.
func runLoopback(t *testing.T, data []byte, blockSize, frameSize, window int, corrupt bool) loopbackResult {
	t.Helper()
	keys, err := DeriveTransferKeys(loopbackMaster())
	if err != nil {
		t.Fatal(err)
	}
	sink := &MemorySink{}

	// Buffers comfortably exceed the window's in-flight frames so no send blocks; the window
	// gate, not the buffer, is what bounds the sender.
	s2r := make(chan []byte, 4096)
	r2s := make(chan []byte, 4096)
	cp := func(f []byte) []byte { return append([]byte(nil), f...) }

	var sentCount int64
	senderSend := func(f []byte) error {
		if corrupt && atomic.AddInt64(&sentCount, 1) == 2 {
			f = cp(f)
			f[len(f)-1] ^= 0x01
		}
		s2r <- cp(f)
		return nil
	}

	sender := NewSender(SenderOptions{
		File:      BytesSource(data, FileMeta{Name: "f", Size: int64(len(data)), Mime: "application/octet-stream", LastModified: 1}, 0),
		Send:      senderSend,
		SendDir:   keys.O2J,
		RecvDir:   keys.J2O,
		BlockSize: blockSize,
		FrameSize: frameSize,
		Window:    window,
	})
	receiver := NewReceiver(ReceiverOptions{
		Send:    func(f []byte) error { r2s <- cp(f); return nil },
		SendDir: keys.J2O,
		RecvDir: keys.O2J,
		Sink:    sink,
	})

	go func() {
		for f := range s2r {
			receiver.Handle(f)
		}
	}()
	go func() {
		for f := range r2s {
			sender.Handle(f)
		}
	}()

	runErrCh := make(chan error, 1)
	go func() { _, e := sender.Run(context.Background()); runErrCh <- e }()

	recvRes, recvErr := receiver.Wait(context.Background())
	var runErr error
	select {
	case runErr = <-runErrCh:
	case <-time.After(5 * time.Second):
		t.Fatal("sender.Run did not return after the receiver settled")
	}
	close(s2r)
	close(r2s)
	return loopbackResult{runErr: runErr, recvRes: recvRes, recvErr: recvErr, sink: sink}
}

func TestLoopbackStreamsVariousSizes(t *testing.T) {
	sizes := []int{0, 1, 7, 8, 20, 4096, 100_000}
	for _, size := range sizes {
		size := size
		t.Run(fmt.Sprintf("%d_bytes", size), func(t *testing.T) {
			data := make([]byte, size)
			for i := range data {
				data[i] = byte((i*131 + 7) & 0xff)
			}
			res := runLoopback(t, data, 1024, 256, 4, false)
			if res.runErr != nil {
				t.Fatalf("sender: %v", res.runErr)
			}
			if res.recvErr != nil {
				t.Fatalf("receiver: %v", res.recvErr)
			}
			if string(res.sink.Bytes()) != string(data) {
				t.Fatalf("received bytes differ from source (%d bytes)", size)
			}
			want := sha256.Sum256(data)
			if res.recvRes.Digest != hex.EncodeToString(want[:]) {
				t.Errorf("digest = %s, want %s", res.recvRes.Digest, hex.EncodeToString(want[:]))
			}
		})
	}
}

// runResumeLoopback wires a resumable sender to a reloaded receiver whose sink and digest are
// pre-seeded with the first haveBlocks blocks, mirroring a real reload from durable storage.
func runResumeLoopback(t *testing.T, data []byte, blockSize, haveBlocks int, senderID, resumeID string) (loopbackResult, int) {
	t.Helper()
	keys, err := DeriveTransferKeys(loopbackMaster())
	if err != nil {
		t.Fatal(err)
	}
	prefix := haveBlocks * blockSize
	if prefix > len(data) {
		prefix = len(data)
	}
	sink := &MemorySink{}
	if resumeID == senderID && prefix > 0 {
		_ = sink.Write(0, data[:prefix])
	}
	seed := NewSHA256Digest()
	if resumeID == senderID {
		seed.Update(data[:prefix])
	}

	s2r := make(chan []byte, 4096)
	r2s := make(chan []byte, 4096)
	cp := func(f []byte) []byte { return append([]byte(nil), f...) }

	sender := NewSender(SenderOptions{
		File:       BytesSource(data, FileMeta{Name: "f", Size: int64(len(data)), Mime: "application/octet-stream", LastModified: 1}, 0),
		Send:       func(f []byte) error { s2r <- cp(f); return nil },
		SendDir:    keys.O2J,
		RecvDir:    keys.J2O,
		BlockSize:  blockSize,
		FrameSize:  256,
		Window:     4,
		TransferID: senderID,
	})
	var commits int
	receiver := NewReceiver(ReceiverOptions{
		Send:       func(f []byte) error { r2s <- cp(f); return nil },
		SendDir:    keys.J2O,
		RecvDir:    keys.O2J,
		Sink:       sink,
		OnProgress: func(int64) { commits++ },
		Resume: &ReceiverResume{
			TransferID: resumeID,
			Files:      map[int]ResumeFileProgress{0: {HaveBlocks: haveBlocks, SeedDigest: seed}},
		},
	})

	go func() {
		for f := range s2r {
			receiver.Handle(f)
		}
	}()
	go func() {
		for f := range r2s {
			sender.Handle(f)
		}
	}()
	runErrCh := make(chan error, 1)
	go func() { _, e := sender.Run(context.Background()); runErrCh <- e }()
	recvRes, recvErr := receiver.Wait(context.Background())
	var runErr error
	select {
	case runErr = <-runErrCh:
	case <-time.After(5 * time.Second):
		t.Fatal("sender.Run did not return after the receiver settled")
	}
	close(s2r)
	close(r2s)
	return loopbackResult{runErr: runErr, recvRes: recvRes, recvErr: recvErr, sink: sink}, commits
}

func TestLoopbackResumesFromHighWaterMark(t *testing.T) {
	blockSize := 1024
	data := make([]byte, 100_000)
	for i := range data {
		data[i] = byte((i*131 + 7) & 0xff)
	}
	blocks := (len(data)-1)/blockSize + 1
	haveBlocks := 10
	id := "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"

	res, commits := runResumeLoopback(t, data, blockSize, haveBlocks, id, id)
	if res.runErr != nil || res.recvErr != nil {
		t.Fatalf("send=%v receive=%v", res.runErr, res.recvErr)
	}
	if string(res.sink.Bytes()) != string(data) {
		t.Fatal("resumed transfer did not reassemble the full file")
	}
	want := sha256.Sum256(data)
	if res.recvRes.Digest != hex.EncodeToString(want[:]) {
		t.Errorf("digest = %s, want %s", res.recvRes.Digest, hex.EncodeToString(want[:]))
	}
	if commits != blocks-haveBlocks {
		t.Errorf("committed %d blocks, want %d (only the missing suffix)", commits, blocks-haveBlocks)
	}
}

func TestLoopbackIgnoresResumeOnTransferIDMismatch(t *testing.T) {
	blockSize := 1024
	data := make([]byte, 20_000)
	for i := range data {
		data[i] = byte((i*131 + 7) & 0xff)
	}
	blocks := (len(data)-1)/blockSize + 1

	// The seed carries a stale id, so the receiver must discard it and receive every block fresh.
	res, commits := runResumeLoopback(t, data, blockSize, 5, "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb", "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
	if res.runErr != nil || res.recvErr != nil {
		t.Fatalf("send=%v receive=%v", res.runErr, res.recvErr)
	}
	if string(res.sink.Bytes()) != string(data) {
		t.Fatal("mismatched-resume transfer did not reassemble the full file")
	}
	if commits != blocks {
		t.Errorf("committed %d blocks, want %d (a full fresh receive)", commits, blocks)
	}
}

func TestLoopbackAbortsOnCorruptedFrame(t *testing.T) {
	data := make([]byte, 4096)
	for i := range data {
		data[i] = byte(i & 0xff)
	}
	res := runLoopback(t, data, 1024, 256, 4, true)
	if res.runErr == nil && res.recvErr == nil {
		t.Fatal("corrupted transfer settled cleanly; expected an abort on at least one side")
	}
	if res.sink.IsClosed() {
		t.Error("sink was closed despite a corrupted frame; it must not commit")
	}
}

type multiMemoryDestination struct {
	manifest Manifest
	sinks    map[string]*MemorySink
	closed   bool
}

func (d *multiMemoryDestination) Prepare(manifest Manifest) error {
	d.manifest = manifest
	d.sinks = make(map[string]*MemorySink, len(manifest.Files))
	return nil
}

func (d *multiMemoryDestination) Open(file FileEntry) (Sink, error) {
	sink := &MemorySink{}
	d.sinks[file.Name] = sink
	return sink, nil
}

func (d *multiMemoryDestination) Close() error { d.closed = true; return nil }

func (d *multiMemoryDestination) Abort(reason string) error {
	for _, sink := range d.sinks {
		_ = sink.Abort(reason)
	}
	return nil
}

func TestLoopbackStreamsNestedMultiFileSet(t *testing.T) {
	keys, err := DeriveTransferKeys(loopbackMaster())
	if err != nil {
		t.Fatal(err)
	}
	contents := [][]byte{{1, 2, 3}, {}, make([]byte, 3000)}
	for i := range contents[2] {
		contents[2][i] = byte(i & 0xff)
	}
	names := []string{"folder/a.bin", "folder/empty.txt", "b.bin"}
	sources := make([]FileSource, len(contents))
	var total int64
	for i, content := range contents {
		sources[i] = BytesSource(content, FileMeta{Name: names[i], Size: int64(len(content)), LastModified: 1}, 0)
		total += int64(len(content))
	}
	s2r := make(chan []byte, 4096)
	r2s := make(chan []byte, 4096)
	copyFrame := func(frame []byte) []byte { return append([]byte(nil), frame...) }
	destination := &multiMemoryDestination{}
	var progress int64
	sender := NewSender(SenderOptions{
		Files: sources, Send: func(frame []byte) error { s2r <- copyFrame(frame); return nil },
		SendDir: keys.O2J, RecvDir: keys.J2O, BlockSize: 1024, FrameSize: 256, Window: 2,
		OnProgress: func(bytes int64) { progress = bytes },
	})
	receiver := NewReceiver(ReceiverOptions{
		Send:    func(frame []byte) error { r2s <- copyFrame(frame); return nil },
		SendDir: keys.J2O, RecvDir: keys.O2J, Destination: destination,
	})
	go func() {
		for frame := range s2r {
			receiver.Handle(frame)
		}
	}()
	go func() {
		for frame := range r2s {
			sender.Handle(frame)
		}
	}()
	sendDone := make(chan struct {
		digest string
		err    error
	}, 1)
	go func() {
		digest, runErr := sender.Run(context.Background())
		sendDone <- struct {
			digest string
			err    error
		}{digest, runErr}
	}()
	received, recvErr := receiver.Wait(context.Background())
	sent := <-sendDone
	close(s2r)
	close(r2s)
	if sent.err != nil || recvErr != nil {
		t.Fatalf("send=%v receive=%v", sent.err, recvErr)
	}
	if sent.digest != received.Digest || progress != total || !destination.closed {
		t.Fatalf("digest/progress/commit mismatch: sent=%s received=%s progress=%d closed=%v",
			sent.digest, received.Digest, progress, destination.closed)
	}
	if len(received.Files) != len(contents) || len(received.Digests) != len(contents) {
		t.Fatalf("received %d files and %d digests", len(received.Files), len(received.Digests))
	}
	for i, name := range names {
		if got := destination.sinks[name].Bytes(); string(got) != string(contents[i]) {
			t.Errorf("%s differs", name)
		}
		want := sha256.Sum256(contents[i])
		if received.Digests[i] != hex.EncodeToString(want[:]) {
			t.Errorf("%s digest = %s", name, received.Digests[i])
		}
	}
}
